File: dart/related.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'models.dart';

// When you have type `Map<String, List<int>>` Dart AOT compiler can't figure
// out that this container can contain only specific type of lists as
// values. However if you create an indirection (like
// `Map<String, TaggedPosts>`) that allows AOT compiler to figure out the type
// of `map[k].posts` because `TaggedPosts` is a concrete class with a single
// implementation - so its static type is enough.
class TaggedPosts {
  final List<int> posts = [];
}

void main() {
  final postsJson =
      jsonDecode(File('../posts.json').readAsStringSync()) as List<dynamic>;

  final posts = postsJson.map(Post.fromJson).toList();

  final sw = Stopwatch()..start();

  final tagMap = <String, TaggedPosts>{};

  for (var i = 0; i < posts.length; i++) {
    for (var tag in posts[i].tags) {
      (tagMap[tag] ??= TaggedPosts()).posts.add(i);
    }
  }

  // preallocate and reuse
  final taggedPostCount = Int32List(posts.length);

  final allRelatedPosts = List.generate(posts.length, (i) {
    final post = posts[i];

    // For now simply inline fillRange which is unfortunately too slow.
    for (var i = 0; i < posts.length; i++) {
      taggedPostCount[i] = 0;
    }

    for (var tag in post.tags) {
      for (var otherPostIdx in tagMap[tag]!.posts) {
        taggedPostCount[otherPostIdx] += 1;
      }
    }

    taggedPostCount[i] = 0; // don't include self

    // Compute max 5 without doing any unnecessary allocations.
    const topN = 5;
    final top5 = Int32List(topN * 2); // flattened list of (count, id)
    var minTags = 0;

    for (var i = 0; i < taggedPostCount.length; i++) {
      final count = taggedPostCount[i];
      if (count <= minTags) {
        continue;
      }

      // Find upper bound: pos at which count is larger than current one.
      var upperBound = (topN - 2) * 2;
      while (upperBound >= 0 && count > top5[upperBound]) {
        top5[upperBound + 2] = top5[upperBound];
        top5[upperBound + 3] = top5[upperBound + 1];
        upperBound -= 2;
      }
      final insertionPos = upperBound + 2;
      top5[insertionPos] = count;
      top5[insertionPos + 1] = i;

      minTags = top5[topN * 2 - 2];
    }

    return {
      "_id": post.iD,
      "tags": post.tags,
      "related": [
        for (var i = 1; i < 10; i += 2) posts[top5[i]],
      ],
    };
  });

  print('Processing time (w/o IO): ${sw.elapsedMilliseconds}ms');

  File('../related_posts_dart.json')
      .writeAsStringSync(jsonEncode(allRelatedPosts));
}

File: ruby/related.rb
# frozen_string_literal: true

require 'json'
require 'time'

TOPN = 5

posts = JSON.parse(File.read('../posts.json'), symbolize_names: true)

t1 = Time.now

POST_SIZE = posts.length
tag_map = Hash.new { |hash, key| hash[key] = [] }

i = 0
while i < POST_SIZE
  tags = posts[i][:tags]
  tag_size = tags&.length.to_i

  j = 0
  while j < tag_size
    tag_map[tags[j]] << i
    j += 1
  end

  i += 1
end

all_related_posts = Array.new(POST_SIZE)
tagged_post_count = Array.new(POST_SIZE, 0)

idx = 0
while idx < POST_SIZE
  post = posts[idx]
  tags = post[:tags]

  tagged_post_count.fill(0)

  ti = 0
  while ti < tags.length
    tag = tags[ti]
    o = tag_map[tag]

    tj = 0
    while tj < o.length
      tagged_post_count[o[tj]] += 1
      tj += 1
    end
    tag_map[tag]

    ti += 1
  end

  tagged_post_count[idx] = 0

  top5 = Array.new(TOPN, 0)
  top_posts = Array.new(TOPN)
  min_tags = 0

  j = 0
  while j < tagged_post_count.length
    count = tagged_post_count[j]
    if count > min_tags
      upper_bound = (TOPN - 2)

      while upper_bound >= 0 && count > top5[upper_bound]
        top5[upper_bound + 1] = top5[upper_bound]
        top_posts[upper_bound + 1] = top_posts[upper_bound]
        upper_bound -= 1
      end

      upper_bound += 1
      top5[upper_bound] = count
      top_posts[upper_bound] = posts[j]

      min_tags = top5[TOPN - 1]
    end
    j += 1
  end

  all_related_posts[idx] = {_id: post[:_id], tags: tags, related: top_posts}
  idx += 1
end

t2 = Time.now

puts "Processing time (w/o IO): #{((t2 - t1) * 1000).round}ms"

File.write('../related_posts_ruby.json', JSON.generate(all_related_posts))
